### PR TITLE
Feature/train model

### PR DESCRIPTION
**Pull request by jakub1090cn**
## Description
This Pull Request includes Training script based on random data.
Training contains callbacks and tensorboard.

## Testing
All tests passed. 

## Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Tests update
- [x] Documentation update
- [x] Changes that do not directly affect the source code or tests

## Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/data.py [hunk 1]
```diff
@@ -99,3 +101,60 @@ def create_data_subsets(df: pd.DataFrame) -> None:
...
+
+        Returns:
+            Tuple[tf.Tensor, tf.Tensor]: A tuple containing a batch of images (X) and labels (y).
+        """
+        indexes = self.indexes[index * self.batch_size:(index + 1) * self.batch_size]
+        X = tf.random.normal(shape=(len(indexes), 224, 224, 3))
+        y = tf.random.uniform(shape=(len(indexes), 1), minval=0, maxval=2, dtype=tf.int32)
+        return X, y
+
+    def on_epoch_end(self):
```
Comment (skrzypczykt) [inline src/data.py]: remove

File: src/data.py [hunk 2]
```diff
@@ -99,3 +101,60 @@ def create_data_subsets(df: pd.DataFrame) -> None:
...
+        """
+        Generates one batch of data.
+
+        Args:
+            index (int): Index of the batch.
+
+        Returns:
+            Tuple[tf.Tensor, tf.Tensor]: A tuple containing a batch of images (X) and labels (y).
+        """
+        indexes = self.indexes[index * self.batch_size:(index + 1) * self.batch_size]
```
Comment (skrzypczykt) [inline src/data.py]: remove

File: src/model_training.py [hunk 3]
```diff
@@ -1,8 +1,49 @@
...
+    Trains the BaldOrNot model using the specified configuration.
+
+    This function initializes the dataset, constructs the model, compiles it with the specified
+    optimizer, loss function, and metrics, and then trains the model on the dataset for the
+    number of epochs defined in the configuration.
+
+    Args:
+        config (BoldOrNotConfig): The configuration object containing model, training, and path parameters.
+    """
+    num_samples = 100
```
Comment (skrzypczykt) [inline src/model_training.py]: move inside mockDataset

File: src/model_training.py [hunk 4]
```diff
@@ -1,8 +1,49 @@
...
+        optimizer=optimizer,
+        loss=config.training_params.loss_function,
+        metrics=config.metrics
+    )
+
+    model.fit(
+        train_dataset,
+        epochs=config.training_params.epochs
+    )
+
```
Comment (skrzypczykt) [inline src/model_training.py]: Callbacks, validation data - same data, history = model.fit, make this function return sth

**Review by skrzypczykt — CHANGES_REQUESTED** [review]

**Comment by jakub1090cn** [comment]
The training script has been updated with the elements mentioned above.

File: tests/test_train_model_integration.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+from src.model_training import train_model
+
+@patch('src.model_training.BaldOrNotModel.fit')
+@patch('src.model_training.BaldDataset')
+def test_train_model(MockBaldDataset, MockFit, test_config):
+    MockFit.return_value = MagicMock()
+
+    history = train_model(test_config)
+
+    MockFit.assert_called_once()
```
Comment (skrzypczykt) [inline tests/test_train_model_integration.py]: Liczba epok zgodna z liczbą linii w history.

Sprawdzenie czy są logi z Tensorboarda,

Czy model sie zapisał?

(ewentualnie) 'EarlyStopping' - sprawdzać czy najlepsze wyniki są z epoki 4 od konca

**Review by skrzypczykt — APPROVED** [review]

File: src/model_training.py [hunk 6]
```diff
@@ -1,53 +1,117 @@
...
+    model.save(model_path)
+    logging.info(f"Model saved at {model_path}")
 
     return history
 
 
+def init_output_dir():
+    project_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    current_date = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+    current_training = f"training{current_date}"
```
Comment (skrzypczykt) [inline src/model_training.py]: move 'training' to arguments

File: scripts/train.py [hunk 7]
```diff
@@ -1,15 +1,30 @@
...
+    output_dir_path = init_output_dir()
+    setup_logging(output_dir_path)
+    config_path = os.path.join(output_dir_path, "config.yaml")
+    shutil.copy("../config.yaml", config_path)
+
     print("Running experiment with the following configuration:")
     print(json.dumps(asdict(config), indent=4))
-    print('Start training:')
-    train_model(config=config)
+    print("Start training:")
```
Comment (skrzypczykt) [inline scripts/train.py]: use logger

File: src/model_training.py [hunk 8]
```diff
@@ -1,53 +1,117 @@
...
     Args:
-        config (BoldOrNotConfig): The configuration object containing model, training, and path parameters.
+        config (BoldOrNotConfig): The configuration object containing model,
+        training, and path parameters.
     """
+    if not check_if_log_exists(output_dir_path):
+        raise RuntimeError(
+            "Log file '{}' not found in '{}'. Make sure logging "
+            "is initialized.".format(LOG_FILE_NAME, output_dir_path)
+        )
```
Comment (skrzypczykt) [inline src/model_training.py]: Create a decorator: https://blog.hubspot.com/website/decorators-in-python
put it in utils

File: src/plot.py [hunk 9]
```diff
@@ -78,3 +91,47 @@ def plot_proportions(
         )
 
     plt.show()
+
+
+def plot_metric_curve(
+    history: History, metric: str, output_dir_path: str
+) -> None:
```
Comment (skrzypczykt) [inline src/plot.py]: rename metric to metric_name

File: src/plot.py [hunk 10]
```diff
@@ -78,3 +91,47 @@ def plot_proportions(
...
+
+    # Save the plot as a PNG file
+    plot_path = os.path.join(output_dir_path, f"{metric}_plot.png")
+    plt.savefig(plot_path)
+    plt.close()
+
+    if not check_if_log_exists(output_dir_path):
+        raise RuntimeError(
+            "Log file '{}' not found in '{}'. Make sure logging "
+            "is initialized.".format(LOG_FILE_NAME, output_dir_path)
```
Comment (skrzypczykt) [inline src/plot.py]: decorate

File: tests/test_train_model_integration.py [hunk 11]
```diff
@@ -1,12 +1,77 @@
...
+    saved correctly in the appropriate directory.
+
+    Args:
+        test_config (BoldOrNotConfig): The model configuration containing the logging parameters.
+        output_dir (str): The output directory path provided by the fixture.
+
+    Raises:
+        AssertionError: If the TensorBoard log directory is not created or if it is empty.
+    """
+    log_dir = "tensorboard_logs_test"
```
Comment (skrzypczykt) [inline tests/test_train_model_integration.py]: log_dir =test_config[...]

File: tests/test_train_model_integration.py [hunk 12]
```diff
@@ -1,12 +1,77 @@
...
+
+    if os.path.exists(log_dir):
+        shutil.rmtree(log_dir)
+
+    train_model(config=test_config, output_dir_path=output_dir)
+
+    assert os.path.exists(log_dir), "Log directory was not created"
+    log_files = os.listdir(log_dir)
+    assert len(log_files) > 0, "Log directory is empty"
```
Comment (skrzypczykt) [inline tests/test_train_model_integration.py]: weird

**Review by skrzypczykt — COMMENTED** [review]
great

**Review by raodz — APPROVED** [review]
lgtm

**Pull request merged**